system/lib/wasmfs/pipe_backend.h

```diff
@@ -0,0 +1,68 @@
+// Copyright 2022 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+#pragma once
+
+#include <queue>
+
+#include "file.h"
+#include "support.h"
+
+namespace wasmfs {
+
+// The data shared between the two sides of a pipe.
+// TODO: something more optimal?
+using PipeData = std::queue<uint8_t>;
+
+// A PipeFile is a simple file that has a reference to a PipeData that it
+// either reads from or writes to. A pair of PipeFiles comprise the two ends of
+// a pipe.
+class PipeFile : public DataFile {
+  std::shared_ptr<PipeData> data;
+
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+    for (size_t i = 0; i < len; i++) {
+      data->push(buf[i]);
+    }
+    return __WASI_ERRNO_SUCCESS;
+  }
+
+  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
+    for (size_t i = 0; i < len; i++) {
+      if (data->empty()) {
+        return __WASI_ERRNO_INVAL;
+      }
```

Comment:
Short reads from pipes are perfectly fine. Is this reported as an error here because short reads are handled at the syscall layer and we assume there won't be any short reads at the backend layer? We should fix that design to allow short reads at the backend layer, too.

Also, I would have expected to see this block and wait for data when the pipe is empty at the beginning of the read. We also need to change the backend API to allow pipes to work correctly: #16269.

Reply:
Yes, the syscall layer will call read only on valid amounts. The read method of File can't return how many bytes it read - it just errors, or not - so it must be called with a proper number, or else it must error.

Not sure what you mean needs fixing there? 

Blocking etc. will be needed eventually, yes. For now this doesn't support the fcntl etc. methods of controlling blocking and all that.

Reply:
The default behavior of pipes is to block on reads of empty pipes, so this is necessary even before fcntl, though.

Reply:
I think that's the normal behavior, but in emscripten we had to change it because blocking isn't possible on the web? Maybe I'm not understanding what change you are proposing here, though - how would we block? Asyncify?

Reply:
Oh, I didn't realize we had different pipe semantics by default. I would have expected either asyncify or just not reading empty pipes on the main thread. Anyway I guess we can iterate on that after this PR.

Reply:
> Not sure what you mean needs fixing there?

Backends like the node backend cannot actually guarantee that short reads won't happen at the backend layer, so we should just allow them. That's also a cleaner design overall because it does not require that extra call to get the file size on each read. But we can fix that in a different PR as well.